File: dbxml/examples/perl/gettingStarted/queryForMetaData.pl

use strict;
use warnings;
use Sleepycat::DbXml 'simple';

sub usage()
{
    warn   <<EOM;
This program retrieves DBXML documents and then retrieves the day and
time the document was stored in the container. The date/time information
was set as metadata information when the document was loaded. See the
exampleLoadContainer.cpp example for how this metadata is set.

Before running this example, use loadExamplesData.[sh|cmd] to pre-load
the appropriate container with documents and metadata.

You are only required to pass this command the path location of the
database environment that you specified when you pre-loaded the
examples data:

    -h <dbenv directory>
EOM

        exit( -1 );
}

sub showTimeStamp
{
    my $mgr = shift ;
    my $container = shift;
    my $query = shift; 
    my $context  = shift;

    # //////////////////////////////////////////////////////////////////////
    # ////  Performs a simple query (with context) against the ///////
    # ////  provided container. The timestamp metadata attribute     ///////
    # ////  value is then displayed.                                 ///////
    # //////////////////////////////////////////////////////////////////////

    # /// some defensive code eliminated for clarity //

    # Used for metadata query
    my $mdURI = "http://dbxmlExamples/timestamp";
    my $mdPrefix = "time";
    my $mdName = "timeStamp";


    #  Perform the query. Result type is by default Result Document
    my $fullQuery = "collection('" . $container->getName() . "')$query";    
    eval {
        print "Exercising query '$fullQuery'\n" ;
        print "Return to continue: ";
        <>;

        my $results = $mgr->query($fullQuery, $context );
        my $value = new XmlValue ;
        while( $results->next($value) )
        {
            my $theDocument = $value->asDocument();

            # Get this document's name
            my $docName= $theDocument->getName() ;

            # Get the timestamp on the document (stored as metadata)
            #  and print it to the console.
            my $metaValue = new XmlValue() ;
            $theDocument->getMetaData( $mdURI, $mdName, $metaValue );
            print "Document '$docName' stored on $metaValue\n";
            #print "Document '$docName' stored on " . $metaValue->asString() . "\n";
            #my $metaValue ;
            #$theDocument->getMetaData( $mdURI, $mdName, $metaValue );
            #print "Document '$docName' stored on $metaValue\n";
        }

        print $results->size() . " objects returned for expression '$fullQuery'\n";

    };
    # Catches XmlException
    if (my $e = catch std::exception)
    {
        warn "Query $fullQuery failed\n";
        warn $e->what() . "\n";
        exit( -1 );
    }
    elsif ($@)
    {
        warn "Query $fullQuery failed\n";
        warn $@;
        exit( -1 );
    }    
}

my $path2DbEnv;
my $theContainer = "namespaceExampleData.dbxml";
while (@ARGV && $ARGV[0] =~ /^-/ )
{
    my $opt = shift @ARGV ;

    if ($opt =~ s/^-h//)
      { $path2DbEnv = $opt || shift }
    else
      { usage() }
}

usage()
     if ! $path2DbEnv ;

 # Open a container in the db environment
my $env = new DbEnv(0);
$env->set_cachesize(0, 64 * 1024 * 1024, 1);
$env->open($path2DbEnv,
    Db::DB_INIT_MPOOL|Db::DB_CREATE|Db::DB_INIT_LOCK|Db::DB_INIT_LOG|Db::DB_INIT_TXN);
my $mgr = new XmlManager($env);

my $containerTxn = $mgr->createTransaction();
my $container = $mgr->openContainer($containerTxn, $theContainer);
$containerTxn->commit();

# Create a context and declare the namespaces
my $context = $mgr->createQueryContext();
$context->setNamespace('fruits'     => 'http://groceryItem.dbxml/fruits');
$context->setNamespace('vegetables' => 'http://groceryItem.dbxml/vegetables');
$context->setNamespace('desserts'   => 'http://groceryItem.dbxml/desserts');

# Find all the vegetables
showTimeStamp( $mgr, $container, "/vegetables:item", $context);

File: dbxml/examples/java/gettingStarted/queryForMetaData.java
//
// See the file LICENSE for redistribution information.
//
// Copyright (c) 2004,2009 Oracle.  All rights reserved.
//
//

package dbxml.gettingStarted;

import java.io.*;
import com.sleepycat.dbxml.*;

class queryForMetaData
{

    private static String theContainer = "namespaceExampleData.dbxml";
    private static String collection = "collection('namespaceExampleData.dbxml')";
    private static void usage() {
	String usageMessage = "\nThis program retrieves DBXML documents and then retrieves the date and time\n";
	usageMessage += "that the document was stored in the container. You should run exampleLoadContainer\n";
	usageMessage += "before running this example. You are only required to pass this command the path\n";
	usageMessage += "location of the database environment that you specified when you loaded the examples\n";
	usageMessage += "data:\n\n";

	usageMessage += "\t-h <dbenv directory>\n";

	usageMessage += "For example:\n";
	usageMessage += "\tjava com.sleepycat.dbxml.examples.gettingStarted.queryForMetaData -h examplesEnvironment\n";

	System.out.println(usageMessage);
	System.exit( -1 );
    }

    //Shows a timestamp for each record that matches the given XPath query.
    // The timestamp is stored as metadata on each document. This metadata was
    // added to the document when the example data was loaded into the container
    // using exampleLoadContainer.java. The timestamp represents the time
    // when the document was loaded into the container.
    private static void showTimeStamp( XmlManager mgr, String query,
				       XmlQueryContext context )
	throws Throwable {
	String fullQuery = collection + query;
	//Perform a single query against the referenced container using
	// the referenced context. The timestamp metadata attribute is then
	// displayed.
	System.out.println("Exercising query: '" + fullQuery + "'.");
	System.out.println("Return to continue: ");
	System.in.read();

	//Perform the query
	XmlResults results = mgr.query(fullQuery, context, null);

	XmlValue value = results.next();
	while (value != null) {
	    //We return the metadata as an XmlValue
	    XmlValue metaValue = new XmlValue();

	    XmlDocument document = value.asDocument();

	    //retrieve the metadata using the uri and name stored in the
	    // mdConst utility class. This uri and name are the same values
	    // used when the metadata was stored on the document.
	    document.getMetaData(mdConst.uri, mdConst.name, metaValue );
	    System.out.println("Document " + document.getName() + " stored on "
			       + metaValue.asString());
	    value = results.next();
	}
	System.out.println(results.size() + " results returned for query '"
			   + fullQuery + "'.");
	results.delete();
    }

    //Utility function to clean up objects, exceptions or not,
    // containers and environments must be closed.
    private static void cleanup(myDbEnv env, XmlContainer openedContainer) {
	try {
	    if (openedContainer != null)
		openedContainer.delete();
	    if (env != null)
		env.cleanup();
	} catch (Exception e) {
	    // ignore exceptions on close
	}
    }

    public static void main(String args[])
	throws Throwable {

	File path2DbEnv = null;
	for(int i = 0; i < args.length; ++i) {
            if (args[i].startsWith("-")) {
		switch(args[i].charAt(1)) {
		case 'h':
		    path2DbEnv = new File( args[++i] );
		    break;
		default:
                    usage();
		}
            }
	}

	if (path2DbEnv == null || ! path2DbEnv.isDirectory()) {
            usage();
	}

	myDbEnv env = null;
	XmlContainer openedContainer = null;
	try {
	    //Open an environment
	    env = new myDbEnv(path2DbEnv);
	    XmlManager theMgr = env.getManager();

	    //Open a container in that environment
	    openedContainer =
		theMgr.openContainer(theContainer);

            //Create a context and declare the namespaces
            XmlQueryContext context = theMgr.createQueryContext();

            context.setNamespace("fruits", "http://groceryItem.dbxml/fruits");
            context.setNamespace("vegetables",
				 "http://groceryItem.dbxml/vegetables");
            context.setNamespace("desserts", "http://groceryItem.dbxml/desserts");

            //Show all the vegetables
            showTimeStamp(theMgr, "/vegetables:item", context);
	} catch (Exception e) {
            System.err.println("Error performing query against " + theContainer );
            System.err.println("   Message: " + e.getMessage() );
            throw e;
	}
	finally {
	    cleanup(env, openedContainer);
	}
    }
} //End main

